ES do not wait for all shards being closed before shutdown/restart

Hi!

I have a one node setup and spotted that if I create a lot of indices/shards and then immediatly will try to restart the server, it will come back online in the red state with a couple of unassigned shards and errors in log like this:

```
Caused by: java.io.EOFException: read past EOF:   
NIOFSIndexInput(path="/var/lib/elasticsearch/elasticsearch/nodes/0/indices/intelligence-reports-2014.05.01/3/index/segments_2”)
```
The problem is that ES starts to flush shards to the hard drive, but does not manage to do that in time, so some shards remain unflushed.

Init script that comes bundled with ES gives 20 seconds for it to finish all preparations:
start-stop-daemon --stop --pidfile "$PID_FILE" \ --user "$ES_USER" \ --retry=TERM/20/KILL/5 > /dev/null

If you bump that period, ES have [hardcoded rule to do that in 30 secs](https://github.com/elastic/elasticsearch/blob/master/src/main/java/org/elasticsearch/indices/IndicesService.java#L153), otherwise:
```
[2015-04-21 15:04:03,558][INFO ][node ] [collaboration] stopping ...   
 [2015-04-21 15:04:33,581][WARN ][indices ] [collaboration] Not all shards are closed yet, waited 30sec - stopping service    
[2015-04-21 15:04:33,581][INFO ][node ] [collaboration] stopped    
[2015-04-21 15:04:33,581][INFO ][node ] [collaboration] closing ...    
[2015-04-21 15:04:33,587][INFO ][node ] [collaboration] closed
```

I know that I can "fix" this by send `flush` after massive write is finished, but why ES needs to kill himself after 30 secs and does not wait for shards to close?

--- a/src/main/java/org/elasticsearch/indices/IndicesService.java
+++ b/src/main/java/org/elasticsearch/indices/IndicesService.java
@@ -96,6 +96,8 @@ import static org.elasticsearch.common.settings.ImmutableSettings.settingsBuilde
  */
 public class IndicesService extends AbstractLifecycleComponent<IndicesService> implements Iterable<IndexService> {
 
+    public static final String INDICES_SHARDS_CLOSED_TIMEOUT = "indices.shards_closed_timeout";
+
     private final InternalIndicesLifecycle indicesLifecycle;
 
     private final IndicesAnalysisService indicesAnalysisService;
@@ -104,6 +106,7 @@ public class IndicesService extends AbstractLifecycleComponent<IndicesService> i
 
     private final PluginsService pluginsService;
     private final NodeEnvironment nodeEnv;
+    private final TimeValue shardsClosedTimeout;
 
     private volatile Map<String, Tuple<IndexService, Injector>> indices = ImmutableMap.of();
     private final Map<Index, List<PendingDelete>> pendingDeletes = new HashMap<>();
@@ -119,6 +122,7 @@ public class IndicesService extends AbstractLifecycleComponent<IndicesService> i
         this.pluginsService = injector.getInstance(PluginsService.class);
         this.indicesLifecycle.addListener(oldShardsStats);
         this.nodeEnv = nodeEnv;
+        this.shardsClosedTimeout = settings.getAsTime(INDICES_SHARDS_CLOSED_TIMEOUT, new TimeValue(1, TimeUnit.DAYS));
     }
 
     @Override
@@ -147,8 +151,8 @@ public class IndicesService extends AbstractLifecycleComponent<IndicesService> i
             });
         }
         try {
-            if (latch.await(30, TimeUnit.SECONDS) == false) {
-              logger.warn("Not all shards are closed yet, waited 30sec - stopping service");
+            if (latch.await(shardsClosedTimeout.seconds(), TimeUnit.SECONDS) == false) {
+              logger.warn("Not all shards are closed yet, waited {}sec - stopping service", shardsClosedTimeout.seconds());
             }
         } catch (InterruptedException e) {
             // ignore

